README.md
# i-can-sort
The software iCanSort is a learning tool to help students study and visualise sorting algorithms and their correctness available on Windows and macOS. This project i-can-sort is built with React and Electron based on NodeJS with npm.

For contributing, please refer to [COUNTRIBUTING.md](./CONTRIBUTING.md).\
For maintenance, please refer to [MaintenanceGuide.md](./MaintenanceGuide.md).\
For code report, please refer to [CodeReport.md](./CodeReport.md).\
For downloading the executable software, please refer to [Release](https://github.com/team10nb/i-can-sort/releases).

# Getting Started with iCanSort

We provide you the installation packages of the software. You can download directly [here](https://github.com/team10nb/i-can-sort/releases).

We also provide you a User Manual [here](https://github.com/team10nb/i-can-sort/blob/main/User%20Manual.pdf).

# Run the Source Code

## 1. Download & Install NodeJS and npm

You’ll need to have `Node@14.15.3` with `npm@6.14.9` but NO later version on your local development machine.

You can follow the link below to download for Windows and Mac. 

For **Windows 64-bit**, you can download and install:

[https://nodejs.org/dist/v14.15.3/node-v14.15.3-x64.msi](https://nodejs.org/dist/v14.15.3/node-v14.15.3-x64.msi)

For **Mac**, you can download and install:

[https://nodejs.org/dist/v14.15.3/node-v14.15.3.pkg](https://nodejs.org/dist/v14.15.3/node-v14.15.3.pkg)

Or you can find downloads that fits your system here by yourself:

 [https://nodejs.org/dist/v14.15.3/](https://nodejs.org/dist/v14.15.3/)

Now you shall be able to execute `node -v` in your terminal and see `v14.15.3`


## 2. Configuration Guide

In the project directory, you should run following command in your terminal in order to configure the dependencies.

### `npm install -g npm@6.14.9`

To install the specific npm version 6.14.9

### `npm -v`

Make sure that you see `6.14.9` in your terminal.
To ensure that you install the specific npm version 6.14.9. 

### `npm install`

Installs all the develepment dependencies.

If this step took more than 10 minutes, you shall consider to connect to eduroam or open a VPN and set proxy for npm.
`npm config set proxy http://server:port`
`npm config set https-proxy http://server:port`

*Note that using a mirror may result in some error for some reason. Please use official registry.*


## 3. Available Script

### `npm start`

Runs the app in the development mode.
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.
You will also see any lint errors in the console.

Now you shall see iCanSort in your bowswer.

*Note that using broswer view the app does not utilise electron features, 'view Group's website' and 'export note' functions cannot work.*

### `npm test`

Launches the test runner in the interactive watch mode.
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Builds the app for production to the `build` folder.
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `npm run electron-start`

Builds the app and launch electron to hava a development view of this app.

### `npm run dist-win`

Builds the app and create an installer for Windows to the `dist` folder.

### `npm run dist-mac`

Builds the app and create an installer for Mac to the `dist` folder.


### `npm run styleguide`

Launch a web view of the Style Guidist.
This will build JavaScript Doc in each folder and display them in a webpage.

# Learn More

We create this app by Create React App. You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).

To learn React, check out the [React documentation](https://reactjs.org/).

To learn Electron, check out the [Electron documentation](https://www.electronjs.org/).








CHANGELOG.md
## [1.0.1](https://github.com/team10nb/i-can-sort/compare/v1.0.0...v1.0.1) (2021-04-08)


### Bug Fixes

* **FirstInBackdrop:** fix compile warning ([e2a82e9](https://github.com/team10nb/i-can-sort/commit/e2a82e9eac23f48b7d79fee7aff217fa44f8d611))
* **Implementation:** fix input format ([181e21b](https://github.com/team10nb/i-can-sort/commit/181e21b9f0c9014f5f4d2066118be2c618d4d146))
* **ModuleProgress:** fix top border radius to 0 ([d9f1255](https://github.com/team10nb/i-can-sort/commit/d9f125562f53a02187cb1fc3115dd0ad8e3bd1e8))
* **PseudoAssertion:** fix code in bubble ([93a1253](https://github.com/team10nb/i-can-sort/commit/93a12539a44ae41bb9f5ce89bb8e566274f4005a))



# [1.0.0](https://github.com/team10nb/i-can-sort/compare/v1.0.0-beta.2...v1.0.0) (2021-04-02)


### Bug Fixes

* **makeStyle:** delete focusVisible in makestyle ([116cf03](https://github.com/team10nb/i-can-sort/commit/116cf03e96bc837f4b58b5deb468fc2fdad0f0be))
* **Termination:** fix bars color ([46182fe](https://github.com/team10nb/i-can-sort/commit/46182feb39effda2f9dd47e1d1a9bba25184d945))


### Features

* **TermHelp:** add term help popup ([f913421](https://github.com/team10nb/i-can-sort/commit/f913421b2a6134912823618b025532372980617e))
* **tutorial term first:** add storage ([a06118e](https://github.com/team10nb/i-can-sort/commit/a06118eea755e8ddc86f99d35982f6192291184b))



# [1.0.0-beta.2](https://github.com/team10nb/i-can-sort/compare/v1.0.0-beta.1...v1.0.0-beta.2) (2021-03-31)


### Bug Fixes

* **Set:** fix link to github page ([2529ba0](https://github.com/team10nb/i-can-sort/commit/2529ba01ea6961e3ce36dd255930bbf86cc66604))
* fix warning in  compile time ([16bb52b](https://github.com/team10nb/i-can-sort/commit/16bb52b76c6013c9c17ca0f0617030e6541b19a7))
* **firstI bug:** fix bug ([7a80312](https://github.com/team10nb/i-can-sort/commit/7a80312b2edd1a9f718effe2774bca8d45ee6211))
* **package.json:** fix windows icon issue ([d8eeea3](https://github.com/team10nb/i-can-sort/commit/d8eeea386bfa6fa8409e816467549a5ecb97f385))


### Features

* **package.json:** update the windows packager script ([f626a6c](https://github.com/team10nb/i-can-sort/commit/f626a6c305688719409dc6047c69f1a8f1129e56))
* **packager:** build and package the software ([c721eeb](https://github.com/team10nb/i-can-sort/commit/c721eeb4f61b905cb3e881d096c7f83ce4a9c7a2))



# [1.0.0-beta.1](https://github.com/team10nb/i-can-sort/compare/17da7f539d6dc52524379ed24dff888e1e2a3ece...v1.0.0-beta.1) (2021-03-29)


### Bug Fixes

* **AnimationCode:** fix issue of pre, div in p ele ([c52464c](https://github.com/team10nb/i-can-sort/commit/c52464ce5a95ca47d67a46077884eb15f0ee22ea))
* **assertion bug:** fix assertion content ([da6d58d](https://github.com/team10nb/i-can-sort/commit/da6d58d9d39e3c0e9ae09ebaf636bc42991d669f))
* **BackDrop:** fix backdrop cannot fill scene ([34314b3](https://github.com/team10nb/i-can-sort/commit/34314b31af1432b4584ce8a1168d04332df7258b))
* **BackDrop:** modify backDrop styles ([9855905](https://github.com/team10nb/i-can-sort/commit/98559058f5138b2225f37a0aff79a98eb34f9e08))
* **buttom name:** change button name ([e93cc99](https://github.com/team10nb/i-can-sort/commit/e93cc99c21c1ce12b19a802eaed41b7dd261d3d3))
* **Correctness:** fix intro and quick sort ([4279b2d](https://github.com/team10nb/i-can-sort/commit/4279b2d9a394c467981bb023db845ed6a13dfe29))
* **Correctness:** fix reset button position issue ([13d189c](https://github.com/team10nb/i-can-sort/commit/13d189ca701185ef8f937dd112ca7af11561ae87))
* **CorrectnessHelp:** fix card size changing issue ([870a5bf](https://github.com/team10nb/i-can-sort/commit/870a5bf0ea9ef6f0b4f190495c75bce278388bb5))
* **CorrectnessProve:** fix function call issue ([6e3c15b](https://github.com/team10nb/i-can-sort/commit/6e3c15bcb65ad2717b8cef204def9ec69f07d801))
* **explanationBar:** fix issue od naming props ([7d206cc](https://github.com/team10nb/i-can-sort/commit/7d206ccad0863945ba4eb9145f0a924fe87fd8be))
* **ExplanationBox:** fix typo of expanation ([5e97222](https://github.com/team10nb/i-can-sort/commit/5e972228a59864bf393cd4cf0c3f07515a793195))
* **ExplanationFrame:** fix consistency problem in size ([b558dc9](https://github.com/team10nb/i-can-sort/commit/b558dc9dbb378e6116fb64238b6764b59fe3eba6))
* **Expression:** fix title merge issue ([fb1ec04](https://github.com/team10nb/i-can-sort/commit/fb1ec04d69f52cc510e1018eb450f93ed69e18ed))
* **FirstIn:** fix snack bar bottom position issue ([3f1fb09](https://github.com/team10nb/i-can-sort/commit/3f1fb0939cabb8840284ae7741789c4ebfd47b6c))
* **FirstIn:** fix text align of welcome ([fe1b9df](https://github.com/team10nb/i-can-sort/commit/fe1b9df4712dec6fd3ecd92d52fe4eadbafa37c4))
* **firstin bug:** fix bug ([cab19bc](https://github.com/team10nb/i-can-sort/commit/cab19bc58056855478af90ba49522561b301bcb4))
* **FirstInBackdrop:** fix borderRadius ([54efd21](https://github.com/team10nb/i-can-sort/commit/54efd2152a12174d0a7c9e508653cfc74b34dcbe))
* **Help:** fix card size changing issue ([386aa0d](https://github.com/team10nb/i-can-sort/commit/386aa0d1f60423a88e085d699a6bdbe404c0510f))
* **Help:** modify help pages of correctness and tutorial ([fb55cac](https://github.com/team10nb/i-can-sort/commit/fb55cac227e01bcd54ce3b0c0dd1ef316d54af5b))
* **Implementation:** change typography to div ([66102c5](https://github.com/team10nb/i-can-sort/commit/66102c53438a18297a6a81fd29ce91d14b94d5db))
* **InputBar:** fix the problem of losing focus ([d014630](https://github.com/team10nb/i-can-sort/commit/d01463057903f4b07df586ab53828334aab5806b))
* **InputTutorial:** fix a typo of capital letter ([8da3e0d](https://github.com/team10nb/i-can-sort/commit/8da3e0d80fd22db2370dcec21e8f3eb0839f4e2b))
* **InputTutorial:** fix firstDayOfWeek issue ([1fe64d7](https://github.com/team10nb/i-can-sort/commit/1fe64d79d3f4dcfd9f8293cf865340152abcd750))
* **Loop:** adjust css ([78d2722](https://github.com/team10nb/i-can-sort/commit/78d27226bf2394d975002072e50ef0ab10c590ee))

main.js
const {app, BrowserWindow} = require('electron')
const path = require('path')
const url = require('url')

let mainWindow

function createWindow () {
mainWindow = new BrowserWindow({
    width: 1050,
    height: 700,
    resizable: false,
    maximizable: false,
    webPreferences: {
      nodeIntegration: true,
      webSecurity: false,
      preload: path.join(__dirname, 'build/preload.js')
    },
  })

mainWindow.loadURL(url.format({
    pathname: path.join(__dirname, 'build/index.html'),
    protocol: 'file:',
    slashes: true
  }))

mainWindow.removeMenu();

mainWindow.setTitle("iCanSort");

mainWindow.on('closed', function () {
    mainWindow = null
  })

mainWindow.on('page-title-updated', (evt) => {
  evt.preventDefault();
});
}

app.on('ready', createWindow)

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', function () {
  if (mainWindow === null) {
    createWindow()
  }
})

app.whenReady().then(() => {
  protocol.interceptFileProtocol('file', (req, callback) => {
    const url = req.url.substr(8);
    callback(slash(decodeURI(url)));
  }, (error) => {
    if (error) {
      console.error('Failed to register protocol');
    }
  });
})



styleguidist.config.js
const path = require('path')

const packagePath = path.resolve(
    __dirname,
    './package.json'
)
const packageFile = require(packagePath)


module.exports = {
  title: "iCanSort",
  version: packageFile.version, 
  usageMode: 'expand', 
  pagePerSection: true, 
  sections: [
    {
      name: 'Components',
      components: './src/components/**/*.jsx',
      ignore: 
      [
        './src/components/**/*.test.jsx',
        './src/components/CorrectnessProof/**',
        './src/components/ProcedureIntroduction/**',
        './src/components/SwitchAnimation/SortingAlgorithms/**',
        './src/components/ProcedureImplementation/**',
      ]
    },
    {
      name: 'Scenes',
      sections:[
        {
          name: 'Tutorial',
          components: 
          [
            './src/scenes/mainPages/Tutorial.jsx',
            './src/scenes/subPages/Tutorial/Swap/Swap.jsx',
            './src/scenes/subPages/Tutorial/Loop/Loop.jsx',
            './src/scenes/subPages/Tutorial/Terminology/Terminology.jsx',
          ]
        },
        {
          name: 'Procedure',
          components: 
          [
            './src/scenes/mainPages/Procedure.jsx',
            './src/scenes/subPages/Procedure/ProcedureSubPage.jsx',
          ]
        },
        {
          name: 'Correctness',
          components: 
          [
            './src/scenes/mainPages/Correctness.jsx',
            './src/scenes/subPages/Correctness/CorrectnessTutorial.jsx',
            './src/scenes/subPages/Correctness/CorrectnessProof.jsx',
          ]
        },
      ]
    }
  ],

  webpackConfig: {
    module: {
      rules: [
        // Babel loader will use your project’s babel.config.js
        {
          test: /\.jsx?$/,
          exclude: /node_modules/,
          loader: 'babel-loader'
        },
        {
          test: /\.css$/,
          use: [
            'style-loader',
            {
              loader: 'css-loader',
              options: {
                importLoaders: 1,
              },
            },
          ],
        },
        {
          test: /\.(woff|woff2|ttf|eot)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
          loader: 'file-loader',
          options: {
            name: '[name].[ext]',
            outputPath: 'assets/fonts',
            publicPath: url => `../assets/fonts/${url}`
          }
        },
        {
          test: /.(jpg|png|gif)$/,
          use:['url-loader']
        },
        {
          test: /\.pdf(\?v=[0-9]\.[0-9]\.[0-9])?$/,
          loader: ['file-loader']
        },  
      ]
    }
  }
}

public/electron.js
const {app, BrowserWindow} = require('electron')
const path = require('path')
const url = require('url')

let mainWindow

function createWindow () {
mainWindow = new BrowserWindow({
    width: 1050,
    height: 700,
    resizable: false,
    maximizable: false,
    webPreferences: {
      nodeIntegration: true,
      webSecurity: false,
      preload: path.join(__dirname, 'preload.js')
    },
    icon: path.join(__dirname, 'icon.png')
  })

mainWindow.loadURL(url.format({
    pathname: path.join(__dirname, 'index.html'),
    protocol: 'file:',
    slashes: true
  }))

mainWindow.removeMenu();

mainWindow.setTitle("iCanSort");

mainWindow.on('closed', function () {
    mainWindow = null
  })

mainWindow.on('page-title-updated', (evt) => {
  evt.preventDefault();
});
}


app.on('ready', createWindow)

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', function () {
  if (mainWindow === null) {
    createWindow()
  }
})

app.whenReady().then(() => {
  protocol.interceptFileProtocol('file', (req, callback) => {
    const url = req.url.substr(8);
    callback(slash(decodeURI(url)));
  }, (error) => {
    if (error) {
      console.error('Failed to register protocol');
    }
  });
})
